### PR TITLE
16 implement exec

### PR DESCRIPTION
**Pull request by VerinAntoine**
- add directory for execution
- add execution header
- add exec function
- add simple execution
- find exec in path #29
- find exec in relative path
- rename file and add exec search in dispatche_cmd
- add exec struct
- replace errors with codes
- fix search pathname join
- add utils founction to child
- update do_exec to use exec struct
- update dispach_cmd to use exec struct
- add doc to duplicate_fds
- add free_exec function
- add call to close and free when error in child #37 #38
- create file path.c
- move free_exec to dispatch.c
- add free_exec to header
- free exec when command is finished
- fix free in dispatch_cmd
- add get_path
- add path.c to sources
- add pathname.c
- move pathname function to pathname.c
- add redirection #48
- add missing header declaration
- norminetted
- update duplicate_fds to take pipe as parameters
- update init_exec to not unset pipes
- finish implementing pipes
- add utils.c
- fix include in exec utils
- move utility functions to exec utils
- add init_exec function
- replace wrong function use in fill_exec
- norminetted #59
- fix stupidity, infile shoud in fact manage infile and not outfile and vice versa
- fix close -1 when file not set
- update documentation
- move pipe close to prevent fd to replace each other in childs
- change main args to include envp
- give envp to execve #51
- add function wait_children
- change return of dispatch_cmd to return exit code
- change dispatch_cmd return codes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -45,6 +46,11 @@ endef
 LIB 		:= $(strip $(LIB))
 
 define SRC 	:=
+	$(addprefix $(EXEC_DIR)/, child.c)
```
Comment (Neffi42) [inline Makefile:49]: group them in one addpreffix plz

File: include/execution.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+/*   Created: 2024/01/11 10:17:31 by averin            #+#    #+#             */
+/*   Updated: 2024/01/23 10:44:54 by averin           ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#ifndef EXECUTION_H
+# define EXECUTION_H
+
+# include "minishell.h"
+# include "errno.h"
```
Comment (Neffi42) [inline include/execution.h:17]: put in inside <>

File: src/execution/child.c [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		(close(exec->outfile), exec->outfile = -1);
+	if (exec->pipes[0] != -1)
+		(close(exec->pipes[0]), exec->pipes[0] = -1);
+	if (exec->pipes[1] != -1)
+		(close(exec->pipes[1]), exec->pipes[1] = -1);
+}
+
+/**
+ * Duplicate fds to stdin and stdout
+ * @param infd fd to duplicate to stdin
```
Comment (Neffi42) [inline src/execution/child.c:33]: wrong doc

File: src/execution/child.c [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	if (exec.outfile != -1)
+		outfd = exec.outfile;
+	else
+		outfd = 1;
+	return (dup2(infd, STDIN_FILENO) == -1
+		|| dup2(outfd, STDOUT_FILENO) == -1);
+}
+
+/**
+ * Execute the command with the args and the env, redirect input and output
```
Comment (Neffi42) [inline src/execution/child.c:55]: doc

File: src/execution/dispatch.c [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	{
+		exec->outfile = open(element, O_WRONLY | O_TRUNC | O_CREAT, 0644);
+		if (exec->outfile == -1)
+			return (perror(element), C_GEN);
+	}
+	return (C_SUCCESS);
+}
+
+/**
+ * Open or create outfile and set corresponding fd in exec
```
Comment (Neffi42) [inline src/execution/dispatch.c:37]: doc, infile and outfile are inverted

File: src/execution/dispatch.c [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+/**
+ * If the command have a pipe create it and store it in exec and close if 
+ * another pipe is already open
+ * @param cmd Where search for pipe
+ * @param exec Where store thie pipe
+ * @return `C_SUCCESS` or `C_GEN` when error
+*/
+static int	init_pipe(t_cmd *cmd, t_exec *exec)
+{
+	if (exec->pipes[0] == -1)
```
Comment (Neffi42) [inline src/execution/dispatch.c:65]: change == to !=

File: src/execution/dispatch.c [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+			if (errno == C_NOEXEC)
+				return (printf("No permission\n"), 127);
+			else if (errno == C_NOFILE)
+				return (printf("Not found\n"), 126);
+		}
+		if (init_infile(cmd, &exec) == C_GEN
+			|| init_outfile(cmd, &exec) == C_GEN)
+			return (125);
+		pid = do_exec(&exec, envp);
+		cmd = find_element(*cmd, T_PIPE);
```
Comment (Neffi42) [inline src/execution/dispatch.c:105]: create a ticket to add bonus types

File: src/execution/path.c [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	i = -1;
+	while (result[++i])
+	{
+		if (result[i][ft_strlen(result[i]) - 1] != '/')
+		{
+			temp = ft_strjoin(result[i], "/");
+			if (!temp)
+				return (ft_fsplit(result), NULL);
+			free(result[i]);
+			result[i] = temp;
```
Comment (Neffi42) [inline src/execution/path.c:37]: one line plz

File: src/execution/utils.c [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+/*   Created: 2024/01/18 15:03:13 by averin            #+#    #+#             */
+/*   Updated: 2024/01/18 15:31:44 by averin           ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "execution.h"
+
+/**
+ * Free execution
+ * TODO: May move to child.c
```
Comment (Neffi42) [inline src/execution/utils.c:17]: doc

File: src/execution/utils.c [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ * @param exec struct to free
+*/
+void	free_exec(t_exec exec)
```
Comment (Neffi42) [inline src/execution/utils.c:20]: doc

File: src/execution/utils.c [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+{
+	free(exec.pathname);
+	ft_fsplit(exec.args);
+}
+
+void	init_exec(t_exec *exec)
```
Comment (Neffi42) [inline src/execution/utils.c:26]: doc

File: src/main.c [hunk 12]
```diff
@@ -6,16 +6,18 @@
...
 /* ************************************************************************** */
 
 #include "minishell.h"
 
-int	main(void)
+int	main(int argc, char **argv, char **envp)
 {
 	char	*line;
 
+	(void)argc;
```
Comment (Neffi42) [inline src/main.c:19]: bruh

**Review by Neffi42 — CHANGES_REQUESTED** [review]

File: src/execution/dispatch.c (hunk 6 above)
Comment (VerinAntoine) [reply src/execution/dispatch.c:65]: won't do fixed in 4c5ba14

**Review by Neffi42 — APPROVED** [review]

**Pull request merged**